[zhoyx] iP

zhoyx opened the pull request:
# Duke
> "Object-oriented programming is an exceptionally bad idea which could only have originated in California. - Edsger Dijkstra [(source)](https://www.brainyquote.com/quotes/edsger_dijkstra_204329)
Duke Frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use
All you need to do is, 

1. Download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. Double-click it.
3. Add your tasks. 
4. Let it manage your tasks for you 😎 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Completed Level 1. Greet, Echo, Exit

zhoyx committed:
Completed Level 2. Add, List

zhoyx committed:
Completed Level 3. Mark as Done

zhoyx committed:
Completed Level 4. ToDos, Events, Deadlines

zhoyx committed:
Modified replies to some erroneous inputs and added comments

zhoyx committed:
Added test cases using I/O redirection

zhoyx committed:
created FormatHelper to add indentation for replies

zhoyx committed:
Updated EXPECTED.TXT

zhoyx committed:
Applied indentation for replies

zhoyx committed:
Introduced Delete feature and improved styling of replies

zhoyx committed:
Updated tests for new reply styling and delete feature

zhoyx committed:
Refactored Duke to extract UI output methods

zhoyx committed:
Created TaskList class and refactored code to uset TaskList methods

zhoyx committed:
Added DukeExceptions to handle Exceptions

zhoyx committed:
Removed accidentally created file

zhoyx committed:
Add methods for file I/O

zhoyx committed:
Add factory method fromSaveFormat to subclasses of Task

zhoyx committed:
Add method to load data from storage

zhoyx committed:
Create data directory if it does not exist

zhoyx committed:
Change how path for data storage is found

zhoyx committed:
Add Date and time feature

zhoyx committed:
Resolve path issue for Storage class

zhoyx committed:
Merge branch 'branch-Level-7'

zhoyx committed:
Resolve merge conflicts

zhoyx committed:
Resolve merge conflict

zhoyx committed:
Merge branch 'add-gradle-support'

zhoyx committed:
Add gradle to project

zhoyx committed:
Add Checkstyle to gradle

jundatan commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    protected static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm");
+    protected LocalDateTime by;
+
+    Deadline(String description, LocalDateTime by) throws IllegalArgumentException {
```
I think it is better to put an access modifier for constructors.

jundatan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,125 @@
+                //Adding an Event
+                int fromPos = currentInput.indexOf(" /from ");
+                int toPos = currentInput.indexOf(" /to ");
+                if (fromPos == -1 || toPos == -1 || toPos > currentInput.length() + 4) {
+                    throw new DukeException("Please include both /from and /to");
+                }
+                if (fromPos > toPos) {
+                    throw new DukeException("Please add the from date first followed by to date");
+                }
+                if (fromPos == 5) {
+                    throw new DukeException("Please include a description of the task");
+                }
+                String description = currentInput.substring(6, fromPos);
+                String from = currentInput.substring(fromPos + 7, toPos);
+                String to = currentInput.substring(toPos+ 5);
+                LocalDateTime convertedFrom = LocalDateTime.parse(from, formatter);
+                LocalDateTime convertedTo = LocalDateTime.parse(to, formatter);
+                taskList.add(new Event(description, convertedFrom, convertedTo));
+            }
+            int count = taskList.size();
+            response.append("  ").append(taskList.getTaskString(count - 1)).append("\n");
+            response.append("Now you have ").append(count).append(" tasks in the list.");
+            memory.saveState(taskList);
+            return response.toString();
+        } catch (DateTimeParseException e) {
+            throw new DukeException("Invalid Date and Time provided, use the format: dd/MM/yyyy HH:mm");
+        }
+    }
+
+    public static String deleteTask(String command) throws DukeException {
+        int index = Integer.parseInt(command.substring(7)) - 1;
+        if (index < 0 || index >= taskList.size()) {
+            throw new DukeException("Error: Please input a valid task index!");
+        } else {
+            String output = taskList.removeTask(index);
+            memory.saveState(taskList);
+            return output;
+        }
+    }
+
```
Perhaps, this should be in the TaskList class instead?

jundatan commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,33 @@
+public abstract class Task {
+    private boolean completed;
```
Maybe isCompleted will be a better naming convention for this boolean variable.

jundatan commented on src/main/java/utils/UI.java:
```diff
@@ -0,0 +1,21 @@
+package utils;
+
+public class UI {
```
Should UI class contain only the greet and signoff print? I think it is better if all the UI related methods are inside this class instead.

jundatan reviewed (commented):
Overall code quality is pretty good. Good OOP but some methods should belong in some classes. Missing header comments but that can be added for better clarity of classes and methods.

zhoyx committed:
Creat Ui class

zhoyx committed:
Move PRINTFORMAT and INPUTFORMAT into FormatHelper

zhoyx committed:
Move TaskList related methods from Duke to TaskList

zhoyx committed:
Organize classies into duke package

zhoyx committed:
Create Command package to abstract away command related methods

zhoyx committed:
Create readCommand in Ui class to get data from user

zhoyx committed:
Implement Parser class

zhoyx committed:
Refactor Duke to implement new classes added

zhoyx committed:
Fix bug on Exit command not printing signoff

zhoyx committed:
Move Task and subclasses into Task package

zhoyx committed:
Create Junit test for Deadline and Todo

zhoyx committed:
Modify mainClassName in build.gradle

zhoyx committed:
Remove unused variable in Duke

zhoyx committed:
Added JavaDocs for Todo

zhoyx committed:
Added JavaDocs for Task

LeonardYam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+            int index = Integer.parseInt(userInput.substring(7)) - 1;
+            return new UnmarkCommand(index);
+        } else if (userInput.matches("^todo .*")) {
+            String description = userInput.substring(5);
+            return new AddTodoCommand(description);
+        } else if (userInput.matches("^deadline .*")) {
+            int byPos = userInput.indexOf(" /by ");
+            if (byPos == -1) {
+                throw new DukeException("Deadline not specified with /by");
+            }
+            String description = userInput.substring(9, byPos);
+            String by = userInput.substring(byPos + 5);
+            LocalDateTime convertedBy = LocalDateTime.parse(by, INPUTFORMAT);
+            return new AddDeadlineCommand(description, convertedBy);
+        } else if (userInput.matches("^event .*")) {
+            int fromPos = userInput.indexOf(" /from ");
+            int toPos = userInput.indexOf(" /to ");
+            if (fromPos == -1 || toPos == -1 || toPos > userInput.length() + 4) {
+                throw new DukeException("Please include both /from and /to");
+            }
+            if (fromPos > toPos) {
+                throw new DukeException("Please add the from date first followed by to date");
+            }
+            if (fromPos == 5) {
+                throw new DukeException("Please include a description of the task");
+            }
+            String description = userInput.substring(6, fromPos);
+            String from = userInput.substring(fromPos + 7, toPos);
+            String to = userInput.substring(toPos + 5);
+            LocalDateTime convertedFrom = LocalDateTime.parse(from, INPUTFORMAT);
+            LocalDateTime convertedTo = LocalDateTime.parse(to, INPUTFORMAT);
+            return new AddEventCommand(description, convertedFrom, convertedTo);
+        } else if (userInput.matches("^delete \\d+")) {
+            int index = Integer.parseInt(userInput.substring(7)) - 1;
+            return new DeleteCommand(index);
+        } else {
+            return new UnknownCommand();
+        }
+    }
+}
```
Perhaps there should be some blank lines between logical blocks to improve readability?

LeonardYam commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,31 @@
+package duke.task;
+
+public abstract class Task {
+    private boolean completed;
```
Should this be prefixed with 'is' to sound more like a boolean?

LeonardYam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+            return new ExitCommand();
+        } else if (userInput.equalsIgnoreCase("list")) {
+            return new ListCommand();
+        } else if (userInput.matches("mark \\d+")) {
+            int index = Integer.parseInt(userInput.substring(5)) - 1;
+            return new MarkCommand(index);
+        } else if (userInput.matches("unmark \\d+")) {
+            int index = Integer.parseInt(userInput.substring(7)) - 1;
+            return new UnmarkCommand(index);
+        } else if (userInput.matches("^todo .*")) {
+            String description = userInput.substring(5);
+            return new AddTodoCommand(description);
+        } else if (userInput.matches("^deadline .*")) {
```
I like the use of regex here!

LeonardYam commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    public int size() {
```
Should this be renamed to start with a verb?

LeonardYam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+package duke;
+
+import duke.command.*;
```
Should the import be more explicit here?

zhoyx committed:
Add JavaDocs for Event

zhoyx committed:
Add JavaDocs for Deadline class

zhoyx committed:
Add JavaDocs for Duke

zhoyx committed:
Add JavaDocs for Parser

zhoyx committed:
Add JavaDocs for Storage

zhoyx committed:
Catch exception thrown for wrong date time input

zhoyx committed:
Add JavaDocs for TaskList

zhoyx committed:
Add JavaDocs for Ui

zhoyx committed:
Add JavaDocs for Command

zhoyx committed:
Add JavaDocs for AddEventCommand, AddDeadlineCommand and AddTodoCommand

zhoyx committed:
Fix Coding standard issues

zhoyx committed:
Implement find method

zhoyx committed:
Merge branch 'branch-A-JavaDoc'

zhoyx committed:
Resolve merge conflicts

zhoyx committed:
Merge branch 'branch-Level-9'

zhoyx committed:
Initialize Javafx

zhoyx committed:
Add image

zhoyx committed:
Refactor Commands to return string and fully implement UI

zhoyx committed:
Add assertions to TaskList and Storage

zhoyx committed:
Fix CheckStyle issues

zhoyx committed:
Add exception for absent Description in Deadline input

zhoyx committed:
Refactor Parser class

zhoyx committed:
Merge pull request #2 from zhoyx/branch-A-Assertions

Add assertions to TaskList and Storage

zhoyx committed:
Merge branch 'master' into branch-A-CodeQuality

zhoyx committed:
Merge pull request #3 from zhoyx/branch-A-CodeQuality

Improve code quality with focus on Parser class

There are several code quality issues, such out of order import statements and 
missing default clause for switch statements, these are resolved to improve readability. 

Refactor Parser class to extract parseEvent and parseDeadline to apply the SLAP 
principle and improve readability.

zhoyx committed:
remove unused TaskList Constructor

zhoyx committed:
Implement "sort deadlines" functionality

Users were only able to see tasks in in the order they were added
using the "list" command.

The "sort deadlines" command now allows users to retrieve only
Deadlines and view them in the sorted order according to their
due date and time.

zhoyx committed:
Create User guide

zhoyx committed:
remove unused singoff in Ui

zhoyx committed:
Merge branch 'branch-Level-10'

zhoyx committed:
Add implementation for bye command

zhoyx committed:
Update readme with bye command

zhoyx committed:
Added more JavaDocs

zhoyx committed:
Remove Dead code in TodoTest

zhoyx committed:
Refactor Load in Storage to reduce nesting

zhoyx committed:
Added new line for readability
